[FE]  공통 컴포넌트인 모달을 만든다.

BadaHertz52 opened the pull request:
- resolves #104

---

### 🚀 어떤 기능을 구현했나요 ?
- 모달을 담아서 createPortal로 root의 바로 아래 형제요소로 띄우는 ModalPortal 구현
- 사이드바를 SideModal로 띄우는 방식으로 수정
- 사이드바의 페이드인, 페이드아웃 동작 오류 수정
- ConfirmModal 구현

### 🔥 어떻게 해결했나요 ?
-  레벨2의 modules 속 모달 구조를 참고했어요.

### 📝 어떤 부분에 집중해서 리뷰해야 할까요?
- 공통 컴포넌트로 만든 모달이라서, 재사용과 확장성을 봐주세요.
- 실제사용 시 사용하는 부분에서 어려운 부분은 없는 지 확인해주세요.

### 📚 참고 자료, 할 말
- 하루 최소 1pr 이거 실화야?

BadaHertz52 committed:
fix: stylelint를 통한 css 속성 정렬 기능 오류 수정

- stylelint 버전16과 충돌되는 플러그인 삭제 : stylelint-config-prettier, stylelint-prettier
- css 정렬에 필요하지 않은 플러그인 삭제 : stylelint-config-standard, stylelint-config-styled-componented, stylelint-webpack-plugin
- 추가로 설치한 플러그인: postcss-syntax, @stylelint/postcss-css-in-js
- stylelint 적용 script 추가
- .stylelintrc.json 수정 : css 관련 rule 설정

BadaHertz52 committed:
refactor: stylelint 적용에 따른 css 속성 정렬

BadaHertz52 committed:
Merge branch 'feat/71-detailed_review_page' of https://github.com/woowacourse-teams/2024-review-me into fe/ci/58-stylelint

BadaHertz52 committed:
fix : 절대 경로 사용 시 오류 수정

오류 : eslintimport/no-unresolved

BadaHertz52 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2024-review-me into fe/fix/79-eslint_import_bug_fix

BadaHertz52 committed:
chore: eslintrc.cjs 에서 불필요한 코드 삭제

 node 환경 setting 삭제

BadaHertz52 committed:
style: eslint 적용에 따른 리뷰 상세페이지 import 순서 정리

BadaHertz52 committed:
refactor: formatDate를 utils/date 파일로 이동

BadaHertz52 committed:
design: theme에 colors,, fontSize 변경 및 borderRadius 추가

BadaHertz52 committed:
feat: MultilineTextViewer 컴포넌트 생성

- 개행이 포함된 string에 개행을 적용해서 보여주는 컴포넌트

BadaHertz52 committed:
feat: 깃허브 저장소 이미지 컴포넌트 생성

BadaHertz52 committed:
feat: 리뷰와 관련된 날짜 UI 컴포넌트 생성

BadaHertz52 committed:
featr: LockButton 삭제 LockToggle 추가

BadaHertz52 committed:
refactor: 피그마 디자인 변경에 따른 ReviewDescription 변경

BadaHertz52 committed:
feat: ReviewComment 컴포넌트 생성

BadaHertz52 committed:
refactor: ReviewViewSection -> ReviewSection 으로 변경 및 리팩토링

- 불필요한  컴포넌트 삭제 : RevieAnswer , ReviewQuestion

BadaHertz52 committed:
refactor: DetailedReviewPage 리팩토링

- 목데이터 변경
-  추가 및 변경된 컴포넌트를 사용해 리뷰 상세페이지 컴포넌트(DetailedReviewPage) 리팩토링
- DetailedReviewPage 폴더의 styles.ts 삭제

BadaHertz52 committed:
refactor: review에 대한 타입 변경

BadaHertz52 committed:
design : ReviewDate의 클론 스타일 적용

BadaHertz52 committed:
feat: KeywordSection 컴포넌트 생성

- 리뷰 상세 페이지 키워드 부분 컴포넌트 생성

BadaHertz52 committed:
feat: ReviewSectionHeader 컴포넌트 생성 및 적용

- 리뷰 상세보기에서 반복되는 질문,키워드 헤더부분을 컴포넌트로 분리

BadaHertz52 committed:
design : 리뷰 상세페이지에 width 변경

BadaHertz52 committed:
refactor: DetailedReview의 목데이터 변경 및 리팩토링

- 타입 변경에 따른 목 데이터 변경
- KeywordSection 적용

BadaHertz52 committed:
design : formWidth를 theme에 추가 및 리뷰 작성/리뷰 상세 페이지에 적용

BadaHertz52 committed:
fix: Layout에서 가로 스크롤 생기는 오류 수정

- 100vw는 스크롤을 포함한 뷰포트 너비라서 100%으로 수정

BadaHertz52 committed:
feat: 리뷰 상페이지 router에 라우터 파라미터 적용 및 관련 설정 변경

- 데모데이를 위해 현재 데이터베이스에 있는 리뷰 상세페이지 id를 sidebar의 리뷰 상세페이지 메뉴 link에 적용
- 리뷰 상세페이지(DetailedReviewPage)의 api 핸들러 수정

BadaHertz52 committed:
docs: 변수명 변경 (isLock -> isPublic)

BadaHertz52 committed:
refactor: 깃헙 저장소 로고 주소 변수명 변경

- projectImgSrc -> thumbnailUrl

BadaHertz52 committed:
ci: msw 관련 패키지 설치

BadaHertz52 committed:
ci: msw 관련 설정파일 추가

- 브라우저 환경, node 환경에서 msw로 목서버 사용할 수 있도록 관련 파일 추가

BadaHertz52 committed:
feat: mock 핸들러 추가 및 상세 리뷰 페이지 목 데이터 추가

BadaHertz52 committed:
feat:  root에서 목서버 사용할 수 있도록함

BadaHertz52 committed:
refactor: endpoint 수정

- env 에서 서버 주소 끝에 슬래시 넣는 것으로 통일

BadaHertz52 committed:
feat: 상세 리뷰 페이지(detailedReviewPage)에 목서버 연결 및 관련 코드 수정

- 상태명 변경: detailReview -> detailedReview
- detailedReview 타입에 null 추가 및 그에 따른 오류 핸들링 추가
- deadline에 string 타입으로 response로 전달되어서 new Date로 감싸서 props로 전달

BadaHertz52 committed:
docs: indexhtml의 title 변경

BadaHertz52 committed:
style: apis/review.ts 의 import 관련 eslint rule 적용에 따른 수정

BadaHertz52 committed:
fix: fetch 오류 수정을 위해 develop pull 받음

BadaHertz52 committed:
fix: ts에서 process 읽지 못하는 오류 수정

BadaHertz52 committed:
fix: webpack dev server script 복원

BadaHertz52 committed:
feat: ModalPortal 셍성

BadaHertz52 committed:
feat: SideModal 컴포넌트, useSide  셍성 및 Sidebar에 적용

BadaHertz52 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2024-review-me into fe/feat/104-common-modal

BadaHertz52 committed:
feat: ModalBackground 컴포넌트 생성 및 적용

BadaHertz52 committed:
fix: 모달 열릴 때 스크롤바 막는 기능 오류 수정

BadaHertz52 committed:
design : ModalPortal 사이즈 단위 변경 (% -> vw, vh)

BadaHertz52 committed:
feat: Button 컴포넌트가 button 속성을 props로 받을 수 있도록 수정

BadaHertz52 committed:
feat: ConfirmModal 생성

BadaHertz52 committed:
refactor: index.tsx에서 모달을 꺼낼 수 있도록 리팩토링

ImxYJL commented on frontend/src/components/common/modals/ConfirmModal/index.tsx:
```diff
@@ -0,0 +1,42 @@
+}
+
+const ConfirmModal: React.FC<PropsWithChildren<ConfirmModalProps>> = ({ children, confirmButton, cancelButton }) => {
+  const buttonList = [confirmButton, cancelButton];
```
각 상황별 모달을 따로 작성하면서도 이렇게 List를 이용해서 나름의 확장성을 가져갈 수 있겠군요~!

chysis commented on frontend/src/components/common/modals/ConfirmModal/index.tsx:
```diff
@@ -0,0 +1,42 @@
+  cancelButton: ConfirmModalButton;
+}
+
+const ConfirmModal: React.FC<PropsWithChildren<ConfirmModalProps>> = ({ children, confirmButton, cancelButton }) => {
```
```suggestion
const ConfirmModal: React.FC<React.PropsWithChildren<ConfirmModalProps>> = ({ children, confirmButton, cancelButton }) => {
```

`React.PropsWithChildren`을 사용하면 불필요한 import를 줄일 수 있을 것 같아요!

chysis commented on frontend/src/components/common/modals/ModalBackground/index.tsx:
```diff
@@ -0,0 +1,9 @@
+
+import * as S from './styles';
+
+function ModalBackground({ children }: PropsWithChildren) {
```
이 부분도 `React.PropsWithChildren`을 사용하면 어떨까요?
괜찮다면 컨벤션 문서에 업데이트해도 좋을 것 같아요.

chysis commented on frontend/src/components/common/modals/ModalPortal/index.tsx:
```diff
@@ -0,0 +1,30 @@
+  id?: string;
+}
+
+const ModalPortal: React.FC<PropsWithChildren<ModalPortalProps>> = ({ children: Modal, id }) => {
```
😊😊

ImxYJL commented on frontend/src/components/common/modals/ModalBackground/index.tsx:
```diff
@@ -0,0 +1,9 @@
+
+import * as S from './styles';
+
+function ModalBackground({ children }: PropsWithChildren) {
```
`PropsWithChildren`을 사용하면 `children` 속성이 optional인데, 지금 `ModalBackground` 컴포넌트에서는 `children`을 거의 필수적으로 받는 컴포넌트라고 생각해서 `children` 속성을 강제로 바꿔주면 어떨까요?
[참고링크!](https://velog.io/@kkojae91/PropsWithChildren%EB%8A%94-%EC%95%88%EC%A0%84%ED%95%9C-%ED%83%80%EC%9E%85%EC%9D%BC%EA%B9%8C)

BadaHertz52 replied:
오!!! PropsWithChildren을 사용하면 children 속성이 optional 이군요!!!
올리 감사해용!

ImxYJL commented on frontend/src/components/common/modals/ModalPortal/index.tsx:
```diff
@@ -0,0 +1,30 @@
+    };
+  });
+
+  return createPortal(<S.ModalPortal id={id || 'modal-portal'}>{Modal}</S.ModalPortal>, document.body);
```
오 저는 그냥 id를 무조건 써줬는데 이렇게 기본값을 주고 따로 열어둘 수도 있겠네요!

chysis commented on frontend/src/hooks/useSidebar.ts:
```diff
@@ -0,0 +1,31 @@
+import { useState } from 'react';
+
+const useSidebar = () => {
+  const CLOSE_TIME = 1000;
+  const OPEN_TIME = 0.5;
```
`CLOSE_TIME`은 ms 단위이고, `OPEN_TIME`은 s 단위인 이유가 있는지 궁금해요.
실제로 동작을 살펴보면 각각 1초, 0.5초로 의도하신 대로 잘 동작하는 것 같네요 :)

추가로, 사용자가 답답함을 느끼지 않을 정도의 짧은 시간으로 어느 정도가 적절할지 함께 이야기 나누면 좋겠네요👍

BadaHertz52 replied:
닫혔을때 오픈보다 더 빨리 닫히게 하고 싶었어요.
시간상 프론트 회의를 열 수 없어서  열리고 닫히는 시간을 저의 취향으로 맞췄어요. 
이 부분은 에프이 말대로 프론트 회의에서 이야기해봐요.

BadaHertz52 replied:
프론트팀과 상의할 시간이 없어서, 열리고 닫히는 시간을 저의 개인적인 취향으로 했어요.
에프이 말대로 프론트팀과 이야기하면 좋을 것 같아요.

chysis commented on frontend/src/components/common/Button/index.tsx:
```diff
@@ -2,7 +2,7 @@ import { ButtonType } from '@/types/styles';
 
 import * as S from './styles';
 
-interface ButtonProps {
+interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
```
👍

chysis reviewed (changes requested):
공통 모달 컴포넌트를 정말 빠르게 만들어주셨네요 👍
사소한 부분에서 코멘트 남겼으니 확인 부탁드려요.
고생하셨어요~

ImxYJL commented on frontend/src/styles/theme.ts:
```diff
@@ -42,13 +42,13 @@ export const colors: ThemeProperty<CSSProperties['color']> = {
 
 export const zIndex: ThemeProperty<CSSProperties['zIndex']> = {
-  sidebar: 999,
+  modal: 999,
```
모달이 공통 컴포넌트라 개별 컴포넌트의 css 파일에서 zindex를 설정하지 않고 theme 파일에 별도로 zIndex를 뒀다고 생각했는데, 제가 생각한 이유가 맞을까요?

BadaHertz52 replied:
zindex는 레이아웃에 영향을 주어서 한 곳에서 관리하는 게 좋다고 생각해서 theme에 만들어두었습니다.😊

ImxYJL reviewed (approved):
너무 멋져부러.

BadaHertz52 committed:
refactor: PropsWithChildren 수정

- PropsWithChildren를 import 하지 않고 React에서 바로 쓸 수 있도록 React.PropsWithChildren로  수정

chysis reviewed (approved):
수정 사항 확인했어요! 고생하셨어요 😊

soosoo22 reviewed (approved):
일단 급하니 머지 할게요~ 추후에 코드 리뷰 진행할게요^^
